### ISSUE TITLE
Negative OSM ID's cause segmentation fault

### ISSUE THREAD

**Issue by chrisfosterelli** [valhalla/valhalla#566]
This is part of the `valhalla_build_tiles` program.

Steps to reproduce:
1. Start with minimal OSM file `test.osm` (see below)
2. Convert file to PBF using `osmconvert test.osm -o=test.pbf` 
3. Run `valhalla_build_tiles --conf /source/conf/valhalla.json test.pbf`
4. Observe segmentation fault immediately (see output below)

Output:
```
root@7a8249db4afe:/converter/ogr2osm# valhalla_build_tiles --conf /source/conf/valhalla.json test.pbf 
2017/02/27 18:16:08.761699 [INFO] Parsing files: test.pbf
2017/02/27 18:16:08.768572 [INFO] Parsing ways...
Segmentation fault
root@7a8249db4afe:/converter/ogr2osm# 
```

Simple XML test file (`test.osm`)
```xml
<?xml version='1.0' encoding='UTF-8'?>
<osm version="0.6">
  <bounds minlat="45.91017" minlon="-64.60505" maxlat="47.89282" maxlon="-61.15594"/>
  <node id="-3" lat="46.2356109" lon="-63.1239977" version="5" timestamp="2014-08-28T13:57:37Z" visible="true"/>
  <node id="-2" lat="46.2351941" lon="-63.1234578" version="5" timestamp="2014-01-19T16:15:21Z" visible="true"/>
  <node id="-1" lat="46.2343808" lon="-63.1223634" version="5" timestamp="2014-01-24T17:08:44Z" visible="true"/>
  <way id="-4" version="14" timestamp="2016-08-28T15:42:43Z">
    <nd ref="-3"/>
    <nd ref="-2"/>
    <nd ref="-1"/>
    <tag k="name" v="Prince Street"/>
    <tag k="highway" v="tertiary"/>
  </way>
</osm>
```

PBF file: [test.pbf.zip](https://github.com/valhalla/valhalla/files/804579/test.pbf.zip)

Running the same file without the negative ID's (and resorted) works in Valhalla without problem. As mentioned in the [OSM docs](http://wiki.openstreetmap.org/wiki/OSM_XML#OSM_XML_file_format_notes), some editor tools may generate negative ID's. 

I found another tool which provides an option to switch these to positive as a functional workaround for my use case, but a more useful error message (or possibly a full solution) may help save other people debugging time and effort.

**Comment by gknisely** [comment]
Is this a editor produced osm file?  Usually, the osm ids are assigned when you upload the data.  

We need the osm ids to properly create the routing graph.  We also require the pbf to be sorted.

**Comment by kevinkreiser** [comment]
we have a little tool to help with `.osm` extracts before converting them to pbf. we could extend [this tool]( scripts/sort_osm.py) to fix up the ids as well

**Comment by chrisfosterelli** [comment]
@gknisely We are using a conversion script to switch from Shapefile to OSM. Most of the tools appear to output negative ID's. [ogr2osm](http://wiki.openstreetmap.org/wiki/Ogr2osm) luckily supports switching back to positive, but this still took a while to track down since Valhalla just segfaults and negative ID's are supposed to be valid :)

The file is sorted, I meant to say that if you switch the ID's in the example file to positive you need to resort since they're no longer sorted the right direction. `osmconvert` warns if you try to convert an unsorted XML file to PBF.

**Comment by kevinkreiser** [comment]
yeah its strange. the only time ive ever encountered negative ids is in osm2pgsql tables where ways which are part of relations get their original id negated to denote they are member of a relation iirc
